sdk/identity/identity/src/credentials/managedIdentityCredential.ts

```diff
@@ -321,6 +321,8 @@ export class ManagedIdentityCredential implements TokenCredential {
         // endpoints are available.  In this case, don't try them in future
         // requests.
         this.isEndpointUnavailable = result === null;
+      } else {
+        throw new CredentialUnavailable("The managed identity endpoint is not currently available");
```

Comment:
Not sure if we want to throw an exception here. Currently, it looks like managed identity returns null rather than throwing to signal "failed because we couldn't reach the endpoint".

Adding the exception may change the existing API, and could be considered a breaking change.

Reply:
I would prefer to throw `CredentialUnavailable` in these cases, otherwise. I think the credential abstraction becomes unclear if sometimes when a credential is unavailable we throw and other times we return null. I fear this results in too many cases which need to handled everywhere credentials are used. However, since this would be a breaking change I am open to debate on this. @bterlson thoughts?

Reply:
I think it is a breaking change. I also agree after talking with Scott that throwing an error makes sense. Would shipping a new major be out of the question for this?

Reply:
Not sure. @ramya-rao-a / @joshfree  - how do we feel about a breaking change/major version bump for aligning how we do error reporting back to the user if a credential fails?

Reply:
I'm in favor of making relatively minor changes like this to get JS inline with the other languages, especially while our product is still very young.

Reply:
@bterlson what's our plan around quirks mode for the JS SDKs?

@jonathandturner why don't we hold off on merging this PR until we have some offline discussions around versioning and compat?

Reply:
@joshfree can you say more about your question re: quirks mode? Not sure I understand that term in this context.

I tend to agree it's very unlikely anyone would be broken by this, but it should also be fairly easy to cut a major if we need.

Reply:
had a conversation with @joshfree and given where we're at with this library, I'm ok taking this technically breaking change because the current behavior is unexpected and is in practice extremely unlikely to break anyone.

Reply:
@bterlson - I'm not sure if we can say it's unexpected if it's in the documentation that it works this way.

"If authentication cannot be performed at this time, this method may return null." from https://docs.microsoft.com/en-us/javascript/api/@azure/identity/managedidentitycredential?view=azure-node-latest

That being said, I agree that throwing an exception would be less surprising that returning null for some cases and throwing an exception in others. After chatting with @schaabs - seems like we can take this fix for the release after this one, which lets us decide what version we want to give it, too.